File: WOT-iOS/WOTPivot/WOTPivot/core/adapters/webRequest/WOTWebProxyRequest.m
//
//  WOTWebProxyRequest.m
//  WOTPivot
//
//  Created by Pavel Yeshchyk on 1/13/20.
//  Copyright © 2020 Pavel Yeshchyk. All rights reserved.
//

#import "WOTWebProxyRequest.h"

@implementation WOTWebProxyRequest

- (NSURLRequest *)finalRequest {
    NSURL *url = [self composedURL];
    NSData *bodyData = self.httpBodyData;
    
    NSMutableURLRequest *request = [[NSMutableURLRequest alloc] initWithURL:url];
    [request setHTTPBody:bodyData];
    [request setTimeoutInterval: 0];
    [request setHTTPMethod:self.method];
    
    NSURLRequest *proxy = [self proxy:request];
    return proxy;
}


- (NSURLRequest *)proxy:(NSURLRequest *)request {
    NSURL *url = request.URL;

    NSURLQueryItem *urlItem = [[NSURLQueryItem alloc] initWithName:@"url" value:url.absoluteString];
    NSURLQueryItem *toItem = [[NSURLQueryItem alloc] initWithName:@"to" value:@"to-txt"];

    NSURLComponents *components = [[NSURLComponents alloc] initWithURL:[NSURL URLWithString:@"https://countwordsfree.com/loadweb"] resolvingAgainstBaseURL:NO];
    [components setQueryItems:@[urlItem, toItem]];

    NSMutableURLRequest *result = [[NSMutableURLRequest alloc] initWithURL:components.URL];
    [result setHTTPMethod:@"POST"];
    return result;
}

- (void)connectionDidFinishLoading:(NSURLConnection *)connection {
    NSData *clearData = [self dataFromProxyData:self.data];
    [self parseData:clearData error:nil];
    self.data = nil;

    [self notifyListenersAboutFinish];
}


- (NSData *)dataFromProxyData:(NSData *)proxyData {

    NSError *serializationError = nil;
    NSString *iso = [[NSString alloc] initWithData:proxyData encoding:NSUTF8StringEncoding];
    NSString *unescaped = [iso stringByReplacingPercentEscapesUsingEncoding:NSUTF8StringEncoding];
    NSData *dutf8 = [unescaped dataUsingEncoding:NSUTF8StringEncoding];
    NSDictionary *jsonData = [NSJSONSerialization JSONObjectWithData:dutf8 options:NSJSONReadingMutableContainers error:&serializationError];
    if (serializationError) {
        return nil;
    }

    NSString *success = jsonData[@"Text"];
    NSData *result = [success dataUsingEncoding:NSUTF8StringEncoding];

    return result;
}

@end

File: WOT-iOS/WOTPivot/WOTPivot/core/adapters/webRequest/WOTWebProxyRequest.swift
//
//  WOTWebProxyRequest.swift
//  WOTPivot
//
//  Created by Pavel Yeshchyk on 3/3/20.
//  Copyright © 2020 Pavel Yeshchyk. All rights reserved.
//

import Foundation

@objc
public class WOTWebProxyRequest: WOTWEBRequest {
    public func dataFrom(proxyData: NSData?) -> NSData? {
        guard let proxyData = proxyData else { return nil }
        var result: Data?
        let iso = String(data: proxyData as Data, encoding: String.Encoding.utf8)
        let unescaped = iso?.removingPercentEncoding
        let dutf8 = unescaped?.data(using: String.Encoding.utf8)
        guard let data = dutf8 else {
            return result as NSData?
        }

        do {
            if let json = try JSONSerialization.jsonObject(with: data, options: JSONSerialization.ReadingOptions.mutableContainers) as? [AnyHashable: Any] {
                if let success = json["Text"] as? String {
                    result = success.data(using: String.Encoding.utf8)
                }
            }

        } catch {}
        return result as NSData?
    }

    public func proxyRequest(for request: NSURLRequest) -> URLRequest? {
        let url = request.url
        let urlItem = URLQueryItem(name: "url", value: url?.absoluteString)
        let toItem = URLQueryItem(name: "to", value: "to-txt")

        guard let proxyURL = URL(string: "https://countwordsfree.com/loadweb") else {
            return nil
        }

        var components = URLComponents(url: proxyURL, resolvingAgainstBaseURL: false)
        components?.queryItems = [urlItem, toItem]
        guard let finalUrl = components?.url else {
            return nil
        }
        var result = URLRequest(url: finalUrl)
        result.httpMethod = "POST"
        return result
    }

//    override public var request: URLRequest {
//        guard let url = self.composedURL() else {
//            fatalError("request not created")
//        }
//        let httpBody = self.httpBodyData
//        let request = NSMutableURLRequest(url: url)
//        request.httpBody = httpBody
//        request.timeoutInterval = 0
//        request.httpMethod = self.method
//        guard let result = self.proxyRequest(for: request) else {
//            fatalError("request not created")
//        }
//        return result
//    }

//    override public func connectionDidFinishLoading(_ connection: NSURLConnection) {
//        let clearData = self.dataFrom(proxyData: self.data as NSData?)
//        self.parse(data: clearData as! Data)
//        self.data = nil
//        self.notifyListenersAboutFinish()
//    }

    /*
     NSData *clearData = [self dataFromProxyData:self.data];
     [self parseData:clearData];
     self.data = nil;

     [self notifyListenersAboutFinish];

     */
}
